lightbeam/validate.py

```diff
@@ -18,19 +33,190 @@ def __init__(self, lightbeam=None):
+                nested_references = self.load_references_structure(swagger, nested_definition)
+                references.update(nested_references)
+        return references
+
+    def load_reference(self, swagger, definition):
+        properties = []
+        if "definitions" in swagger.keys():
+            schema = swagger["definitions"][definition]
+        elif "components" in swagger.keys() and "schemas" in swagger["components"].keys():
+            schema = swagger["components"]["schemas"][definition]
+        else:
+            self.logger.critical(f"Swagger contains neither `definitions` nor `components.schemas` - check that the Swagger is valid.")
+        for k in schema["properties"].keys():
+            if k in schema.get("required",[]):
+                properties.append(k)
+        return properties
+
+    def get_swagger_definition_for_endpoint(self, endpoint):
+        return util.camel_case(self.lightbeam.config["namespace"]) + "_" + util.singularize_endpoint(endpoint)
+    
     # Validates a single endpoint based on the Swagger docs
-    def validate_endpoint(self, swagger, endpoint, local_descriptors=[]):
-        definition = util.camel_case(self.lightbeam.config["namespace"]) + "_" + util.singularize_endpoint(endpoint)
+    async def validate_endpoint(self, endpoint):
+        fail_fast_threshold = self.lightbeam.config.get("validate",{}).get("references",{}).get("max_failures", 10)
+        definition = self.get_swagger_definition_for_endpoint(endpoint)
+        data_files = self.lightbeam.get_data_files_for_endpoint(endpoint)
+        tasks = []
+        total_counter = 0
+        for file_name in data_files:
+            self.logger.info(f"validating {file_name} against {definition} schema...")
+            with open(file_name) as file:
+                self.lightbeam.num_errors = 0
+                for line_counter, line in enumerate(file):
+                    total_counter += 1
+                    data = line.strip()
+                        
+                    tasks.append(asyncio.create_task(
+                        self.do_validate_payload(endpoint, file_name, data, line_counter)))
+                
+                    if total_counter%self.MAX_VALIDATE_TASK_QUEUE_SIZE==0:
+                        await self.lightbeam.do_tasks(tasks, total_counter, log_status_counts=False)
+                        tasks = []
+                        if total_counter%1000==0:
+                            self.logger.info(f"(processed {total_counter}...)")
+                    
+                    # implement "fail fast" feature:
+                    if self.lightbeam.num_errors >= fail_fast_threshold:
+                        self.logger.critical(f"... STOPPING; found {self.lightbeam.num_errors} >= validate.references.max_failures={fail_fast_threshold} VALIDATION ERRORS.")
+                        break
+
+            if len(tasks)>0: await self.lightbeam.do_tasks(tasks, total_counter, log_status_counts=False)
```

Comment:
Confirmed that the case I've identified can happen:
```
...
2024-07-02 11:18:59.928 lightbeam WARNING ... VALIDATION ERROR (line 8): payload contains an invalid assessmentReference : {"assessmentIdentifier": "COGO", "namespace": "uri://www.riversideinsights.com/cognitive_abilities_test"}
2024-07-02 11:19:00.377 lightbeam WARNING ... VALIDATION ERROR (line 9): payload contains an invalid studentReference : {"studentUniqueId": "6548888888"}
2024-07-02 11:19:49.831 lightbeam CRITICAL ... STOPPING; found 90 >= validate.references.max_failures=10 VALIDATION ERRORS.
```
Notice the 49-second gap between the 10th error and the stoppage

Reply:
> Suppose there are many validation tasks and all of them fail. Currently, the job won't be aborted until MAX_VALIDATE_TASK_QUEUE_SIZE tasks have completed."

You're right. The `Lightbeam` class already has state about how many errors have happened so far. I've pushed a small change that makes do_validate_payload() first check that fewer then fail_fast_threshold errors have happened so far, otherwise it returns immediately. This means that the rest of the task queue will still process, but incredibly quickly, and not really do anything.

Regarding
> we have this extra line that does the "leftover" tasks, and it would be nice to not have to remember to do this

I actually considered something like this in the past, the issue is that if you're processing a _huge_ JSONL file with millions of rows, the `tasks` queue will get huge and possibly use up Python's memory. That's the reason to do it in batches as they're read from the file, to be less memory-hungry. I'm definitely open to another way to implement this, though, if you can think of one.

Reply:
> You're right. The Lightbeam class already has state about how many errors have happened so far. I've pushed a small change that makes do_validate_payload() first check that fewer then fail_fast_threshold errors have happened so far, otherwise it returns immediately. This means that the rest of the task queue will still process, but incredibly quickly, and not really do anything.

Simple and effective, I like it!

> I actually considered something like this in the past, the issue is that if you're processing a huge JSONL file with millions of rows, the tasks queue will get huge and possibly use up Python's memory. That's the reason to do it in batches as they're read from the file, to be less memory-hungry. I'm definitely open to another way to implement this, though, if you can think of one.

Mmm, good point. I'll put this in the back of my mind for now. Seems like a more general challenge that we'll perhaps think of a solution to someday.